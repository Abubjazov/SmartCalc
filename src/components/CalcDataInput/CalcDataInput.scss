@import '../../variables';

.calc-data-input {
	display: grid;
	grid-template-columns: 300px;
	grid-template-rows: 50px 38px 38px;
	grid-row: auto;
	row-gap: 10px;
	justify-items: center;

	.add-item {
		justify-self: start;
		width: 30px;
		height: 30px;
		padding-left: 6px;
		padding-top: 1px;
		margin-left: 13px;
		background-color: $main-color;
		color: $white;
		font-size: 30px;
		line-height: 27px;
		border-radius: 50px;
		cursor: pointer;

		&:hover {
			box-shadow: 1px 1px 10px $grey;
		}
	}

	.added-item {
		margin-bottom: 13px;
		display: flex;
	}

	.remove-item {
		margin-left: 10px;
		width: 24px;
		height: 24px;
		padding-left: 7px;
		padding-top: 0;
		background-color: $red;
		color: $white;
		font-size: 30px;
		line-height: 19px;
		border-radius: 50px;
		cursor: pointer;

		&:hover {
			box-shadow: 1px 1px 10px $main-color;
		}
	}

	button {
		margin-top: 30px;
		border: none;
		border-radius: 20px;
		width: 100px;
		height: 40px;
		text-align: center;
		font-size: 18px;
		line-height: 18px;
		background-color: $main-color;
		color: $white;
		transition: all 0.1s linear;
		cursor: pointer;

		&:hover {
			box-shadow: 1px 1px 10px $grey;
		}
	}

	input {
		justify-self: start;
	}
}
