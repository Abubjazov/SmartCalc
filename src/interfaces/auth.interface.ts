export enum AuthActionTypes {
	FETCH_TOKEN = 'FETCH_TOKEN',
	FETCH_TOKEN_SUCCESS = 'FETCH_TOKEN_SUCCESS',
	FETCH_TOKEN_ERROR = 'FETCH_TOKEN_ERROR',

	DELETE_TOKEN = 'DELETE_TOKEN',
	DELETE_TOKEN_SUCCESS = 'DELETE_TOKEN_SUCCESS',
	DELETE_TOKEN_ERROR = 'DELETE_TOKEN_ERROR',

	CLEAR_ERROR = 'CLEAR_ERROR',
}

export interface AuthData {
	email: string
	password: string
}

export interface UserData {
	token: string
	name: string
	email: string
}

export interface AuthState {
	token: null | string
	name: string
	email: string
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchTokenAction {
	type: AuthActionTypes.FETCH_TOKEN
	payload: AuthData
}

interface FetchTokenSuccessAction {
	type: AuthActionTypes.FETCH_TOKEN_SUCCESS
	payload: UserData
}

interface FetchTokenErrorAction {
	type: AuthActionTypes.FETCH_TOKEN_ERROR
	payload: string
}

interface DeleteTokenAction {
	type: AuthActionTypes.DELETE_TOKEN
}

interface DeleteTokenSuccessAction {
	type: AuthActionTypes.DELETE_TOKEN_SUCCESS
}

interface DeleteTokenErrorAction {
	type: AuthActionTypes.DELETE_TOKEN_ERROR
	payload: string
}

interface ClearErrorAction {
	type: AuthActionTypes.CLEAR_ERROR
}

export type AuthAction =
	| FetchTokenAction
	| FetchTokenSuccessAction
	| FetchTokenErrorAction
	| DeleteTokenAction
	| DeleteTokenSuccessAction
	| DeleteTokenErrorAction
	| ClearErrorAction
