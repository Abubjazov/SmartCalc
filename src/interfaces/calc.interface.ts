export enum CalcActionTypes {
	FETCH_CURRENT_STATE = 'FETCH_CURRENT_STATE',
	FETCH_CURRENT_STATE_SUCCESS = 'FETCH_CURRENT_STATE_SUCCESS',
	FETCH_CURRENT_STATE_ERROR = 'FETCH_CURRENT_STATE_ERROR',
}

export interface InputItems {
	key: string
	value: string
}

export interface UserState {
	items: number[]
	step: number
}

export interface CalcState {
	items: number[]
	inputItems: InputItems[]
	sortedItems: number[]
	sortDirection: boolean
	searchString: string
	status: 'waiting' | 'error' | 'loading'
	step: number
	error: null | string
}

interface FetchCurrentStateAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE
}

interface FetchCurrentStateSuccessAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE_SUCCESS
	payload: UserState
}

interface FetchCurrentStateErrorAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE_ERROR
	payload: string
}

export type CalcAction =
	| FetchCurrentStateAction
	| FetchCurrentStateSuccessAction
	| FetchCurrentStateErrorAction
