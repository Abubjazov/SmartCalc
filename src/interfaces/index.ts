export * from './auth.interface'
export * from './calc.interface'
