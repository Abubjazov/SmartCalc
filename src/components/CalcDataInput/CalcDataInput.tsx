import { nanoid } from 'nanoid'
import { Helmet } from 'react-helmet'

import { useActions } from '../../hooks/useActions'
import { useTypedSelector } from '../../hooks/useTypedSelector'
import { InputItem } from '../../interfaces'

import './CalcDataInput.scss'

export const CalcDataInput = (): JSX.Element => {
	const { inputItems } = useTypedSelector(state => state.calc)
	const { token } = useTypedSelector(state => state.auth)
	const { addInputItem, changeInputItem, removeInputItem, switchToConfirm } =
		useActions()

	const convertArray = (arr: InputItem[]): number[] =>
		arr.map(item => +item.value)

	return (
		<div className='calc-data-input'>
			<Helmet>
				<title>Ввод данных</title>
			</Helmet>
			<h1>Введите данные</h1>
			{inputItems.map((item, index) =>
				index < 2 ? (
					<input
						key={item.key}
						type='number'
						placeholder='Введите число'
						onChange={e => changeInputItem(item.key, e.target.value)}
						value={item.value}
					/>
				) : (
					<div className='added-item' key={item.key}>
						<input
							type='number'
							placeholder='Введите число'
							onChange={e => changeInputItem(item.key, e.target.value)}
							value={item.value}
						/>
						<div
							className='remove-item'
							onClick={() => removeInputItem(item.key)}
						>
							-
						</div>
					</div>
				)
			)}

			{inputItems.length < 10 ? (
				<div
					className='add-item'
					onClick={() => addInputItem({ key: nanoid(), value: '' })}
				>
					+
				</div>
			) : null}

			<button onClick={() => switchToConfirm(convertArray(inputItems), token)}>
				Далее
			</button>
		</div>
	)
}
