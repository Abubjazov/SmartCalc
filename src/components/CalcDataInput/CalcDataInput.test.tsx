import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import { CalcActionTypes } from '../../interfaces'
import { CalcDataInput } from './CalcDataInput'

const mockStore = configureStore()

describe('Component: CalcDataInput', () => {
	test('should render CalcDataInput by default', () => {
		const store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataInput with added items < 10', () => {
		const store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '3' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render CalcDataInput with added items > 10', () => {
		const store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '3' },
					{ key: 'a13key4', value: '4' },
					{ key: 'a13key5', value: '5' },
					{ key: 'a13key6', value: '6' },
					{ key: 'a13key7', value: '7' },
					{ key: 'a13key8', value: '8' },
					{ key: 'a13key9', value: '9' },
					{ key: 'a13key10', value: '10' },
					{ key: 'a13key11', value: '11' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const { asFragment } = render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should dispatch an action on "Remove item" Button Click', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '3' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions = [
			{ type: CalcActionTypes.REMOVE_INPUT_ITEM, payload: 'a13key3' },
		]

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		userEvent.click(screen.getByText('-'))
		expect(store.getActions()).toEqual(expectedActions)
	})

	test('should dispatch an action on "Add item" Button Click', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions = CalcActionTypes.ADD_INPUT_ITEM

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		userEvent.click(screen.getByText('+'))
		expect(store.getActions()[0].type).toEqual(expectedActions)
	})

	test('should dispatch an action on "Remove item" Key Down', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
					{ key: 'a13key3', value: '3' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions = [
			{ type: CalcActionTypes.REMOVE_INPUT_ITEM, payload: 'a13key3' },
			{ type: CalcActionTypes.REMOVE_INPUT_ITEM, payload: 'a13key3' },
		]

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		fireEvent.keyDown(screen.getByText('-'), { key: 'Space', code: 'Space' })
		fireEvent.keyDown(screen.getByText('-'), { key: 'Enter', code: 'Enter' })
		fireEvent.keyDown(screen.getByText('-'), {
			key: 'ArrowDown',
			code: 'ArrowDown',
		})
		expect(store.getActions()).toEqual(expectedActions)
	})

	test('should dispatch an action on "Add item" Key Down', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions = CalcActionTypes.ADD_INPUT_ITEM

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		fireEvent.keyDown(screen.getByText('+'), { key: 'Space', code: 'Space' })
		expect(store.getActions()[0].type).toEqual(expectedActions)

		fireEvent.keyDown(screen.getByText('+'), { key: 'Enter', code: 'Enter' })
		expect(store.getActions()[1].type).toEqual(expectedActions)

		fireEvent.keyDown(screen.getByText('+'), {
			key: 'ArrowDown',
			code: 'ArrowDown',
		})
		expect(store.getActions().length).toEqual(2)
	})

	test('should dispatch an action on "Next" Button Click, SubmitStatus is TRUE', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '1' },
					{ key: 'a13key2', value: '2' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions = [
			{
				type: CalcActionTypes.SWITCH_TO_CONFIRM,
				payload: { items: [1, 2], token: 'testtoken' },
			},
		]

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		userEvent.click(screen.getByText('Далее'))
		expect(store.getActions()).toEqual(expectedActions)
	})

	test('should dispatch an action on "Next" Button Click, SubmitStatus is FALSE', () => {
		const middlewares = [thunk]
		const mockStore = configureStore(middlewares)

		let store = mockStore({
			calc: {
				inputItems: [
					{ key: 'a13key1', value: '' },
					{ key: 'a13key2', value: '' },
				],
			},
			auth: {
				token: 'testtoken',
			},
		})

		const expectedActions: {
			type: CalcActionTypes
			payload: {
				items: number[]
				token: string
			}
		}[] = []

		render(
			<Provider store={store}>
				<CalcDataInput />
			</Provider>
		)

		userEvent.click(screen.getByText('Далее'))
		expect(store.getActions()).toEqual(expectedActions)
	})
})
