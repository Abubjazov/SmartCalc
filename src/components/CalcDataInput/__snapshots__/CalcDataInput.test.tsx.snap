// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalcDataInput should render CalcDataInput by default 1`] = `
<DocumentFragment>
  <div
    class="calc-data-input"
  >
    <h1>
      Введите данные
    </h1>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value=""
      />
    </div>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value=""
      />
    </div>
    <div
      class="add-item"
      tabindex="0"
    >
      +
    </div>
    <button>
      Далее
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: CalcDataInput should render CalcDataInput with added items < 10 1`] = `
<DocumentFragment>
  <div
    class="calc-data-input"
  >
    <h1>
      Введите данные
    </h1>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value="1"
      />
    </div>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value="2"
      />
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="3"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="add-item"
      tabindex="0"
    >
      +
    </div>
    <button>
      Далее
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: CalcDataInput should render CalcDataInput with added items > 10 1`] = `
<DocumentFragment>
  <div
    class="calc-data-input"
  >
    <h1>
      Введите данные
    </h1>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value="1"
      />
    </div>
    <div
      class="calc-input"
    >
      <input
        class="null"
        placeholder="Введите число"
        type="text"
        value="2"
      />
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="3"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="4"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="5"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="6"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="7"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="8"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="9"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="10"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <div
      class="added-item"
    >
      <div
        class="calc-input"
      >
        <input
          class="null"
          placeholder="Введите число"
          type="text"
          value="11"
        />
      </div>
      <div
        class="remove-item"
        tabindex="0"
      >
        -
      </div>
    </div>
    <button>
      Далее
    </button>
  </div>
</DocumentFragment>
`;
